<?xml version="1.0" ?>

<robot name="r2d2">
    <material name="white">
        <color rgba="1 1 1 1"/>
    </material>
    <material name="black">
        <color rgba="0 0 0 1"/>
    </material>
    <material name="blue">
        <color rgba="0 0 .8 1"/>
    </material>


    <link name="body">
        <visual>
            <geometry>
                <cylinder length=".6" radius=".2"/>
            </geometry>
            <material name="blue"/>
        </visual>
    </link>


    <link name="right_leg">
        <visual>
            <geometry>
                <box size=".6 .2 .1"/>
            </geometry>
            <origin xyz="0 0 -0.3" rpy="0 1.5707 0"/>
            <material name="white"/>
        </visual>
    </link>

    <joint name="body_to_right_leg" type="fixed">
        <parent link="body"/>
        <child link="right_leg"/>
        <!-- origin relative to parent's frame -->
        <origin xyz=".22 0 .25"/>
    </joint>

    <link name="right_foot">
        <visual>
            <geometry>
                <box size=".1 .4 .1"/>
            </geometry>
            <material name="white"/>
        </visual>
    </link>

    <joint name="right_leg_to_foot" type="fixed">
        <parent link="right_leg"/>
        <child link="right_foot"/>
        <origin xyz="0 0 -.6"/>
    </joint>

    <link name="right_front_wheel">
        <visual>
            <geometry>
                <cylinder length=".1" radius=".035"/>
            </geometry>
            <material name="black"/>
            <origin rpy="0 1.5707 0"/>
        </visual>
    </link>

    <joint name="right_foot_to_front" type="fixed">
        <parent link="right_foot"/>
        <child link="right_front_wheel"/>
        <origin xyz="0 .13 -.05"/>
    </joint>

    <link name="right_rear_wheel">
        <visual>
            <geometry>
                <cylinder length=".1" radius=".035"/>
            </geometry>
            <material name="black"/>
            <origin rpy="0 1.5707 0"/>
        </visual>
    </link>

    <joint name="right_foot_to_rear" type="fixed">
        <parent link="right_foot"/>
        <child link="right_rear_wheel"/>
        <origin xyz="0 -.13 -.05"/>
    </joint>


    <link name="left_leg">
        <visual>
            <geometry>
                <box size=".6 .2 .1"/>
            </geometry>
            <origin xyz="0 0 -0.3" rpy="0 1.5707 0"/>
            <material name="white"/>
        </visual>
    </link>

    <joint name="body_to_left_leg" type="fixed">
        <parent link="body"/>
        <child link="left_leg"/>
        <!-- origin relative to parent's frame -->
        <origin xyz="-.22 0 .25"/>
    </joint>

    <link name="left_foot">
        <visual>
            <geometry>
                <box size=".1 .4 .1"/>
            </geometry>
            <material name="white"/>
        </visual>
    </link>

    <joint name="left_leg_to_foot" type="fixed">
        <parent link="left_leg"/>
        <child link="left_foot"/>
        <origin xyz="0 0 -.6"/>
    </joint>

    <link name="left_front_wheel">
        <visual>
            <geometry>
                <cylinder length=".1" radius=".035"/>
            </geometry>
            <material name="black"/>
            <origin rpy="0 1.5707 0"/>
        </visual>
    </link>

    <joint name="left_foot_to_front" type="fixed">
        <parent link="left_foot"/>
        <child link="left_front_wheel"/>
        <origin xyz="0 .13 -.05"/>
    </joint>

    <link name="left_rear_wheel">
        <visual>
            <geometry>
                <cylinder length=".1" radius=".035"/>
            </geometry>
            <material name="black"/>
            <origin rpy="0 1.5707 0"/>
        </visual>
    </link>

    <joint name="left_foot_to_rear" type="fixed">
        <parent link="left_foot"/>
        <child link="left_rear_wheel"/>
        <origin xyz="0 -.13 -.05"/>
    </joint>


    <link name="head">
        <visual>
            <geometry>
                <sphere radius=".2"/>
            </geometry>
            <material name="white"/>
        </visual>
    </link>

    <joint name="head_swivel" type="continuous">
        <parent link="body"/>
        <child link="head"/>
        <origin xyz="0 0 .3"/>
        <axis xyz="0 0 1"/>
    </joint>

    <link name="eye_pole">
        <visual>
            <geometry>
                <cylinder length="0.3" radius="0.02"/>
            </geometry>
            <material name="white"/>
            <origin xyz="0 0 -.15"/>
        </visual>
    </link>

    <joint name="periscope" type="prismatic">
        <parent link="head"/>
        <child link="eye_pole"/>
        <origin xyz="-.04 .1414 .1414" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit lower="0" upper=".29" effort="10" velocity=".5"/>
    </joint>

    <link name="eye">
        <visual>
            <geometry>
                <box size=".08 .08 .08"/>
            </geometry>
            <material name="blue"/>
        </visual>
    </link>

    <joint name="eye_pole_to_eye" type="fixed">
        <parent link="eye_pole"/>
        <child link="eye"/>
    </joint>
</robot>
